cartographer/mapping/2d/tsdf_2d.cc

```diff
@@ -0,0 +1,176 @@
+  if (*tsdf_cell >= value_converter->getUpdateMarker()) {
+    return;
+  }
+  mutable_update_indices()->push_back(flat_index);
+  mutable_known_cells_box()->extend(cell_index.matrix());
+  *tsdf_cell =
+      value_converter->TSDToValue(tsd) + value_converter->getUpdateMarker();
+  *weight_cell = value_converter->WeightToValue(weight);
+  return;
+}
+
+float TSDF2D::GetTSD(const Eigen::Array2i& cell_index) const {
+  if (limits().Contains(cell_index)) {
+    return value_converter->ValueToTSD(
+        correspondence_cost_cells()[ToFlatIndex(cell_index)]);
+  }
+  return value_converter->getMinTSD();
+}
+
+float TSDF2D::GetWeight(const Eigen::Array2i& cell_index) const {
+  if (limits().Contains(cell_index)) {
+    return value_converter->ValueToWeight(
+        weight_cells_[ToFlatIndex(cell_index)]);
+  }
+  return value_converter->getMinWeight();
+}
+
+std::pair<float, float> TSDF2D::GetTSDAndWeight(
+    const Eigen::Array2i& cell_index) const {
+  if (limits().Contains(cell_index)) {
+    int flat_index = ToFlatIndex(cell_index);
+    return std::make_pair(
+        value_converter->ValueToTSD(correspondence_cost_cells()[flat_index]),
+        value_converter->ValueToWeight(weight_cells_[flat_index]));
+  }
+  return std::make_pair(value_converter->getMinTSD(),
+                        value_converter->getMinWeight());
+}
+
+void TSDF2D::GrowLimits(const Eigen::Vector2f& point) {
+  Grid2D::GrowLimits(point,
+                     {mutable_correspondence_cost_cells(), &weight_cells_},
+                     {value_converter->getUnknownTSDValue(),
+                      value_converter->getUnknownWeightValue()});
+}
+
+proto::Grid2D TSDF2D::ToProto() const {
+  proto::Grid2D result;
+  result = Grid2D::ToProto();
+  result.mutable_tsdf_2d()->mutable_weight_cells()->Reserve(
+      weight_cells_.size());
+  for (const auto& cell : weight_cells_) {
+    result.mutable_tsdf_2d()->mutable_weight_cells()->Add(cell);
+  }
+  result.mutable_tsdf_2d()->set_truncation_distance(
+      value_converter->getMaxTSD());
+  result.mutable_tsdf_2d()->set_max_weight(value_converter->getMaxWeight());
+  return result;
+}
+
+std::unique_ptr<Grid2D> TSDF2D::ComputeCroppedGrid() const {
+  Eigen::Array2i offset;
+  CellLimits cell_limits;
+  ComputeCroppedLimits(&offset, &cell_limits);
+  const double resolution = limits().resolution();
+  const Eigen::Vector2d max =
+      limits().max() - resolution * Eigen::Vector2d(offset.y(), offset.x());
+  std::unique_ptr<TSDF2D> cropped_grid = common::make_unique<TSDF2D>(
+      MapLimits(resolution, max, cell_limits), value_converter->getMaxTSD(),
+      value_converter->getMaxWeight());
+  for (const Eigen::Array2i& xy_index : XYIndexRangeIterator(cell_limits)) {
+    if (!IsKnown(xy_index + offset)) continue;
+    cropped_grid->SetCell(xy_index, GetTSD(xy_index + offset),
+                          GetWeight(xy_index + offset));
+  }
+  return std::unique_ptr<Grid2D>(cropped_grid.release());
```

Comment:
just return cropped_grid instead - since it's derived from Grid2D that will/should work.